File: Services/MousePerformanceEventBuffer.vb
Imports WpfApp1.Models

Namespace Services
    Friend NotInheritable Class MousePerformanceEventBuffer
        Private Const ChunkSize As Integer = 2048

        Private ReadOnly _chunks As New List(Of MousePerformanceEvent())()
        Private _count As Integer

        Public ReadOnly Property Count As Integer
            Get
                Return _count
            End Get
        End Property

        Public Sub Clear()
            _chunks.Clear()
            _count = 0
        End Sub

        Public Sub Add(item As MousePerformanceEvent)
            If item Is Nothing Then
                Return
            End If

            Dim chunkIndex = _count \ ChunkSize
            Dim offset = _count Mod ChunkSize
            If chunkIndex >= _chunks.Count Then
                _chunks.Add(New MousePerformanceEvent(ChunkSize - 1) {})
            End If

            _chunks(chunkIndex)(offset) = item
            _count += 1
        End Sub

        Default Public ReadOnly Property Item(index As Integer) As MousePerformanceEvent
            Get
                If index < 0 OrElse index >= _count Then
                    Throw New ArgumentOutOfRangeException(NameOf(index))
                End If

                Return _chunks(index \ ChunkSize)(index Mod ChunkSize)
            End Get
        End Property

        Public Function CreateReadOnlyView(snapshotCount As Integer) As IReadOnlyList(Of MousePerformanceEvent)
            Dim clampedCount = Math.Max(0, Math.Min(snapshotCount, _count))
            Dim chunkSnapshot = _chunks.ToArray()
            Return New MousePerformanceEventReadOnlyView(chunkSnapshot, clampedCount)
        End Function

        Private NotInheritable Class MousePerformanceEventReadOnlyView
            Implements IReadOnlyList(Of MousePerformanceEvent)

            Private ReadOnly _chunks As MousePerformanceEvent()()
            Private ReadOnly _count As Integer

            Public Sub New(chunks As MousePerformanceEvent()(), count As Integer)
                _chunks = If(chunks, Array.Empty(Of MousePerformanceEvent())())
                _count = count
            End Sub

            Default Public ReadOnly Property Item(index As Integer) As MousePerformanceEvent Implements IReadOnlyList(Of MousePerformanceEvent).Item
                Get
                    If index < 0 OrElse index >= _count Then
                        Throw New ArgumentOutOfRangeException(NameOf(index))
                    End If

                    Return ResolveItem(index)
                End Get
            End Property

            Public ReadOnly Property Count As Integer Implements IReadOnlyCollection(Of MousePerformanceEvent).Count
                Get
                    Return _count
                End Get
            End Property

            Public Function GetEnumerator() As IEnumerator(Of MousePerformanceEvent) Implements IEnumerable(Of MousePerformanceEvent).GetEnumerator
                Return New Enumerator(_chunks, _count)
            End Function

            Private Function GetUntypedEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
                Return GetEnumerator()
            End Function

            Private Function ResolveItem(index As Integer) As MousePerformanceEvent
                Dim chunkIndex = index \ ChunkSize
                Dim offset = index Mod ChunkSize
                If chunkIndex < 0 OrElse chunkIndex >= _chunks.Length OrElse _chunks(chunkIndex) Is Nothing Then
                    Throw New InvalidOperationException("The event buffer snapshot is not available.")
                End If

                Return _chunks(chunkIndex)(offset)
            End Function

            Private NotInheritable Class Enumerator
                Implements IEnumerator(Of MousePerformanceEvent)

                Private ReadOnly _chunks As MousePerformanceEvent()()
                Private ReadOnly _count As Integer
                Private _index As Integer

                Public Sub New(chunks As MousePerformanceEvent()(), count As Integer)
                    _chunks = If(chunks, Array.Empty(Of MousePerformanceEvent())())
                    _count = count
                    _index = -1
                End Sub

                Public ReadOnly Property Current As MousePerformanceEvent Implements IEnumerator(Of MousePerformanceEvent).Current
                    Get
                    If _index < 0 OrElse _index >= _count Then
                        Throw New InvalidOperationException()
                    End If

                    Dim chunkIndex = _index \ ChunkSize
                    Dim offset = _index Mod ChunkSize
                    If chunkIndex < 0 OrElse chunkIndex >= _chunks.Length OrElse _chunks(chunkIndex) Is Nothing Then
                        Throw New InvalidOperationException("The event buffer snapshot is not available.")
                    End If

                    Return _chunks(chunkIndex)(offset)
                End Get
            End Property

                Private ReadOnly Property UntypedCurrent As Object Implements IEnumerator.Current
                    Get
                        Return Current
                    End Get
                End Property

                Public Function MoveNext() As Boolean Implements IEnumerator.MoveNext
                    If _index >= _count Then
                        Return False
                    End If

                    _index += 1
                    Return _index < _count
                End Function

                Public Sub Reset() Implements IEnumerator.Reset
                    _index = -1
                End Sub

                Public Sub Dispose() Implements IDisposable.Dispose
                End Sub
            End Class
        End Class
    End Class
End Namespace

File: src/ClickSyncMouseTester/Services/MousePerformanceEventBuffer.cs
using ClickSyncMouseTester.Models;
using System;
using System.Collections;
using System.Collections.Generic;

namespace ClickSyncMouseTester.Services;

internal sealed class MousePerformanceEventBuffer
{
    private sealed class MousePerformanceEventReadOnlyView : IReadOnlyList<MousePerformanceEvent>
    {
        private sealed class Enumerator : IEnumerator<MousePerformanceEvent>
        {
            private readonly MousePerformanceEvent[][] _chunks;

            private readonly int _count;

            private int _index;

            public MousePerformanceEvent Current
            {
                get
                {
                    if (_index < 0 || _index >= _count)
                    {
                        throw new InvalidOperationException();
                    }
                    int chunkIndex = _index / ChunkSize;
                    int offsetInChunk = _index % ChunkSize;
                    if (chunkIndex < 0 || chunkIndex >= _chunks.Length || _chunks[chunkIndex] == null)
                    {
                        throw new InvalidOperationException("The event buffer snapshot is not available.");
                    }
                    return _chunks[chunkIndex][offsetInChunk];
                }
            }

            object IEnumerator.Current => Current;

            public Enumerator(MousePerformanceEvent[][] chunks, int count)
            {
                _chunks = chunks ?? Array.Empty<MousePerformanceEvent[]>();
                _count = count;
                _index = -1;
            }

            public bool MoveNext()
            {
                if (_index >= _count)
                {
                    return false;
                }
                _index++;
                return _index < _count;
            }

            bool IEnumerator.MoveNext()
            {
                return this.MoveNext();
            }

            public void Reset()
            {
                _index = -1;
            }

            void IEnumerator.Reset()
            {
                this.Reset();
            }

            public void Dispose()
            {
            }

            void IDisposable.Dispose()
            {
                this.Dispose();
            }
        }

        private readonly MousePerformanceEvent[][] _chunks;

        private readonly int _count;

        public MousePerformanceEvent this[int index]
        {
            get
            {
                if (index < 0 || index >= _count)
                {
                    throw new ArgumentOutOfRangeException("index");
                }
                return ResolveItem(index);
            }
        }

        public int Count => _count;

        public MousePerformanceEventReadOnlyView(MousePerformanceEvent[][] chunks, int count)
        {
            _chunks = chunks ?? Array.Empty<MousePerformanceEvent[]>();
            _count = count;
        }

        public IEnumerator<MousePerformanceEvent> GetEnumerator()
        {
            return new Enumerator(_chunks, _count);
        }

        IEnumerator<MousePerformanceEvent> IEnumerable<MousePerformanceEvent>.GetEnumerator()
        {
            return this.GetEnumerator();
        }

        private IEnumerator GetUntypedEnumerator()
        {
            return GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetUntypedEnumerator();
        }

        private MousePerformanceEvent ResolveItem(int index)
        {
            int chunkIndex = index / ChunkSize;
            int offsetInChunk = index % ChunkSize;
            if (chunkIndex < 0 || chunkIndex >= _chunks.Length || _chunks[chunkIndex] == null)
            {
                throw new InvalidOperationException("The event buffer snapshot is not available.");
            }
            return _chunks[chunkIndex][offsetInChunk];
        }
    }

    private const int ChunkSize = 2048;

    private readonly List<MousePerformanceEvent[]> _chunks;

    private int _count;

    public int Count => _count;

    public MousePerformanceEvent this[int index]
    {
        get
        {
            if (index < 0 || index >= _count)
            {
                throw new ArgumentOutOfRangeException("index");
            }
            return _chunks[index / ChunkSize][index % ChunkSize];
        }
    }

    public MousePerformanceEventBuffer()
    {
        _chunks = new List<MousePerformanceEvent[]>();
    }

    public void Clear()
    {
        _chunks.Clear();
        _count = 0;
    }

    public void Add(MousePerformanceEvent mouseEvent)
    {
        if (mouseEvent == null)
        {
            return;
        }

        int chunkIndex = _count / ChunkSize;
        int offsetInChunk = _count % ChunkSize;
        if (chunkIndex >= _chunks.Count)
        {
            _chunks.Add(new MousePerformanceEvent[ChunkSize]);
        }
        _chunks[chunkIndex][offsetInChunk] = mouseEvent;
        _count++;
    }

    public IReadOnlyList<MousePerformanceEvent> CreateReadOnlyView(int snapshotCount)
    {
        int count = Math.Max(0, Math.Min(snapshotCount, _count));
        return new MousePerformanceEventReadOnlyView(_chunks.ToArray(), count);
    }
}






